File: backend/app/services/ingest.py
from datetime import datetime
from typing import Iterable

from fastapi import HTTPException, status
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.security import verify_api_key
from app.models import Agent, LogEntry
from app.schemas import IngestPayload, LogEvent


async def _get_agent(session: AsyncSession, agent_id: str) -> Agent | None:
    result = await session.execute(select(Agent).where(Agent.id == agent_id))
    return result.scalar_one_or_none()


def _prepare_rows(events: Iterable[LogEvent], received_at: datetime) -> list[LogEntry]:
    return [
        LogEntry(
            agent_id=event.agent_id,
            log_source=event.log_source,
            event_time=event.event_time,
            received_at=received_at,
            event_level=event.event_level,
            event_id=event.event_id,
            message=event.message,
        )
        for event in events
    ]


async def ingest_events(session: AsyncSession, payload: IngestPayload, api_key: str) -> None:
    agent = await _get_agent(session, payload.agent_id)
    if not agent or not verify_api_key(api_key, agent.api_key_hash):
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Invalid credentials")

    for event in payload.events:
        if event.agent_id != payload.agent_id:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Agent mismatch")

    received_at = datetime.utcnow()
    rows = _prepare_rows(payload.events, received_at)

    session.add_all(rows)
    await session.commit()

File: backend/src/routes/ingest.ts
import { FastifyInstance, FastifyReply, FastifyRequest } from "fastify";

import { pool } from "../db.js";
import { verifyApiKey } from "../security.js";
import { ingestPayloadSchema } from "../types.js";

interface IngestRequest extends FastifyRequest {
  Headers: {
    "x-agent-key"?: string;
  };
}

export const registerIngestRoutes = async (app: FastifyInstance) => {
  app.post("/logs/ingest", async (request: IngestRequest, reply: FastifyReply) => {
    const apiKey = request.headers["x-agent-key"];
    if (!apiKey) {
      return reply.status(401).send({ error: "Missing API key" });
    }

    const parseResult = ingestPayloadSchema.safeParse(request.body);
    if (!parseResult.success) {
      return reply.status(400).send({ error: "Invalid payload", details: parseResult.error.flatten() });
    }

    const payload = parseResult.data;

    const mismatch = payload.events.find((event) => event.agent_id !== payload.agent_id);
    if (mismatch) {
      return reply.status(400).send({ error: "Agent mismatch" });
    }

    const agentResult = await pool.query(
      "SELECT id, api_key_hash FROM agents WHERE id = $1",
      [payload.agent_id]
    );

    const agent = agentResult.rows[0];
    if (!agent || !verifyApiKey(apiKey, agent.api_key_hash)) {
      return reply.status(403).send({ error: "Invalid credentials" });
    }

    const values: Array<string | number> = [];
    const placeholders: string[] = [];

    payload.events.forEach((event, index) => {
      const baseIndex = index * 7;
      placeholders.push(`($${baseIndex + 1}, $${baseIndex + 2}, $${baseIndex + 3}, $${baseIndex + 4}, $${baseIndex + 5}, $${baseIndex + 6}, $${baseIndex + 7})`);
      values.push(
        event.agent_id,
        event.log_source,
        event.event_time,
        event.event_level.toUpperCase(),
        event.event_id,
        event.message,
        new Date().toISOString()
      );
    });

    if (values.length > 0) {
      await pool.query(
        `INSERT INTO logs (agent_id, log_source, event_time, event_level, event_id, message, received_at)
         VALUES ${placeholders.join(", ")}`,
        values
      );
    }

    return reply.status(202).send({ status: "accepted", received: payload.events.length });
  });
};
